File: src/main/java/io/binac/leetcode/SearchInRotatedSortedArray.java
package io.binac.leetcode;

/**
 * Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.
 * <p>
 * <p>(i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).
 * <p>
 * <p>You are given a target value to search. If found in the array return its index, otherwise return -1.
 * <p>
 * <p>You may assume no duplicate exists in the array.
 * <p>
 * <p>Your algorithm's runtime complexity must be in the order of O(log n).
 * <p>
 * <p>Example 1:
 * <blockquote><pre>
 *     Input: nums = [4,5,6,7,0,1,2], target = 0
 *     Output: 4
 * </blockquote></pre>
 * Example 2:
 * <blockquote><pre>
 *     Input: nums = [4,5,6,7,0,1,2], target = 3
 *     Output: -1
 * </blockquote></pre>
 */
public class SearchInRotatedSortedArray {
    public static class Solution1 {
        public int search(int[] nums, int target) {
            int low = 0, high = nums.length - 1;
            while (low <= high) {
                int mid = (low + high) >>> 1;
                if (nums[mid] == target) {
                    return mid;
                } else if (nums[mid] < target) {
                    if (nums[low] <= nums[mid]) {
                        low = mid + 1;
                    } else {
                        if (nums[high] >= target)
                            low = mid + 1;
                        else
                            high = mid - 1;
                    }
                } else {
                    if (nums[mid] <= nums[high]) {
                        high = mid - 1;
                    } else {
                        if (nums[low] <= target)
                            high = mid - 1;
                        else
                            low = mid + 1;
                    }
                }
            }
            return -1;
        }
    }
}

File: python3/leetcodepy/search_in_rotated_sorted_array.py
"""
There is an integer array nums sorted in ascending order (with distinct values).

Prior to being passed to your function, nums is rotated at an unknown pivot index k (0 <= k < nums.length) such that the resulting array is [nums[k], nums[k+1], ..., nums[n-1], nums[0], nums[1], ..., nums[k-1]] (0-indexed). For example, [0,1,2,4,5,6,7] might be rotated at pivot index 3 and become [4,5,6,7,0,1,2].

Given the array nums after the rotation and an integer target, return the index of target if it is in nums, or -1 if it is not in nums.



Example 1:

Input: nums = [4,5,6,7,0,1,2], target = 0
Output: 4

Example 2:

Input: nums = [4,5,6,7,0,1,2], target = 3
Output: -1

Example 3:

Input: nums = [1], target = 0
Output: -1


Constraints:

1 <= nums.length <= 5000
-10^4 <= nums[i] <= 10^4
All values of nums are unique.
nums is guaranteed to be rotated at some pivot.
-10^4 <= target <= 10^4


Follow up: Can you achieve this in O(log n) time complexity?
"""
from typing import List


class Solution1:
    def search(self, nums: List[int], target: int) -> int:
        low, high = 0, len(nums) - 1
        while low <= high:
            mid = (low + high) // 2
            val = nums[mid]
            if val == target:
                return mid
            if nums[low] <= nums[mid]:
                if nums[low] <= target < nums[mid]:
                    high = mid - 1
                else:
                    low = mid + 1
            else:
                if nums[mid] < target <= nums[high]:
                    low = mid + 1
                else:
                    high = mid - 1
        return -1
